File: solutions/zig/luhn/1/luhn.zig
const std = @import("std");

pub fn isValid(s: []const u8) bool {
    var i: usize = s.len;
    var sum: usize = 0;
    var count: usize = 0;
    var double = false;
    while (i > 0) {
        i -= 1;
        const char = s[i];
        if (char == ' ') {
            continue;
        }
        if (char < '0' or char > '9') {
            return false;
        }
        if (double) {
            const doubled = (char - '0') * 2;
            sum += if (doubled > 9) doubled - 9 else doubled;
        } else {
            sum += char - '0';
        }
        double = !double;
        count += 1;
    }
    return (count > 1) and (sum % 10 == 0);
}

File: solutions/php/luhn/1/Luhn.php
<?php
declare(strict_types=1);

function isValid(string $number): bool
{
    $i = strlen($number);
    $sum = 0;
    $count = 0;
    $double = false;
    while ($i > 0) {
        $i--;
        $char = $number[$i];
        if ($char === ' ') {
            continue;
        }
        if ($char < '0' || $char > '9') {
            return false;
        }
        $digit = ord($char) - ord('0');
        if ($double) {
            $doubled = $digit * 2;
            $sum += $doubled > 9 ? $doubled - 9 : $doubled;
        } else {
            $sum += $digit;
        }
        $double = !$double;
        $count++;
    }
    return $count > 1 && $sum % 10 === 0;
}
